components/core/tests/test-end_to_end.cpp

```diff
@@ -0,0 +1,127 @@
+    auto const default_compression_level = 3;
+    auto const default_print_archive_stats = false;
+    auto const default_ordered = false;
+    auto const default_target_ordered_chunk_size = 0;
+    auto structurize_arrays = GENERATE(true, false);
+
+    Cleanup const test_cleanup;
+
+    std::filesystem::create_directory(cTestEndToEndArchiveDirectory);
+    REQUIRE(std::filesystem::is_directory(cTestEndToEndArchiveDirectory));
+
+    clp_s::JsonParserOption parser_option{};
+    parser_option.file_paths.push_back(get_test_input_local_path());
+    parser_option.archives_dir = cTestEndToEndArchiveDirectory;
+    parser_option.target_encoded_size = default_target_encoded_size;
+    parser_option.max_document_size = default_max_document_size;
+    parser_option.min_table_size = default_min_table_size;
+    parser_option.compression_level = default_compression_level;
+    parser_option.print_archive_stats = default_print_archive_stats;
+    parser_option.structurize_arrays = structurize_arrays;
+
+    clp_s::JsonParser parser{parser_option};
+    REQUIRE(parser.parse());
+    parser.store();
+
+    REQUIRE(false == std::filesystem::is_empty(cTestEndToEndArchiveDirectory));
+
+    std::filesystem::create_directory(cTestEndToEndOutputDirectory);
+    REQUIRE(std::filesystem::is_directory(cTestEndToEndOutputDirectory));
+
+    clp_s::JsonConstructorOption constructor_option{};
+    constructor_option.archives_dir = parser_option.archives_dir;
+    constructor_option.output_dir = cTestEndToEndOutputDirectory;
+    constructor_option.ordered = false;
+    constructor_option.target_ordered_chunk_size = 0;
+    for (auto const& entry : std::filesystem::directory_iterator(constructor_option.archives_dir)) {
+        if (false == entry.is_directory()) {
+            // Skip non-directories
+            continue;
+        }
+
+        constructor_option.archive_id = entry.path().filename();
+        clp_s::JsonConstructor constructor{constructor_option};
+        constructor.store();
+    }
+
+    std::filesystem::path extracted_json_path{cTestEndToEndOutputDirectory};
+    extracted_json_path /= "original";
+    REQUIRE(std::filesystem::exists(extracted_json_path));
+
+    int result = std::system("command -v jq >/dev/null 2>&1");
+    REQUIRE(0 == result);
+    std::string command = fmt::format(
+            "jq --sort-keys --compact-output '.' {}/original | sort > {}",
+            cTestEndToEndOutputDirectory,
+            cTestEndToEndOutputSortedJson
+    );
+    result = std::system(command.c_str());
+    REQUIRE(0 == result);
+
+    REQUIRE(false == std::filesystem::is_empty(cTestEndToEndOutputSortedJson));
+
+    result = std::system("command -v diff >/dev/null 2>&1");
+    REQUIRE(0 == result);
+    command = fmt::format(
+            "diff -u {} {}  > /dev/null",
+            cTestEndToEndOutputSortedJson,
+            get_test_input_local_path()
+    );
+    result = std::system(command.c_str());
+    REQUIRE(0 == WEXITSTATUS(result));
```

Comment:
The [docs](https://pubs.opengroup.org/onlinepubs/9699919799/functions/wait.html) for `WEXITSTATUS` suggest that it's only valid to use `WEXITSTATUS` if we check `WIFEXITED` first, right?